Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,640 @@
+    UNIMPLEMENTED();
+    return StubReturn();
+}
+
+/**
+ @Status Stub
+ @Notes
+*/
+void NSDecimalRound(NSDecimal* result, const NSDecimal* number, NSInteger scale, NSRoundingMode roundingMode) {
+    UNIMPLEMENTED();
+}
+
+static void _InternalDecimalCopy(NSDecimal* destination, NSDecimalInternal* source) {
+    if (!source || !destination) {
+        return;
+    }
+    NSDecimalCopy(destination, &(source->decimal));
+    memcpy(destination->_mantissa, source->_mantissa, NSDecimalMaxSize * sizeof(unsigned short));
+}
+
+static NSComparisonResult _compareMantissa(NSDecimal* left, NSDecimal* right) {
+    // Note: All inputs have been normalized via NSDecimalNormalize
+    if (left->_length > right->_length) {
+        return NSOrderedDescending;
+    }
+
+    if (left->_length < right->_length) {
+        return NSOrderedAscending;
+    }
+
+    // same length
+    for (int index = right->_length - 1; index >= 0; index--) {
+        if (right->_mantissa[index] > left->_mantissa[index]) {
+            return NSOrderedAscending;
+        }
+
+        if (right->_mantissa[index] < left->_mantissa[index]) {
+            return NSOrderedDescending;
+        }
+    }
+
+    return NSOrderedSame;
+}
+
+static NSCalculationError _handleDifferenceOfSignAddition(NSDecimal* result,
+                                                          NSDecimal* left,
+                                                          NSDecimal* right,
+                                                          NSRoundingMode roundingMode) {
+    // Note: All inputs have been normalized via NSDecimalNormalize
+    // Since these are normalized and to account for unsigned subtraction we have to do mantissa based comparison.
+
+    NSComparisonResult res = _compareMantissa(left, right);
+
+    if (res == NSOrderedSame) {
+        SET_TO_ZERO(result);
+        return NSCalculationNoError;
+    }
+
+    if (res == NSOrderedAscending) {
+        // R > L
+        NSCalculationError error = _subtraction(result, right, left);
+        if (error != NSCalculationNoError) {
+            INVALIDATE_DECIMAL(result);
+            return error;
+        }
+        result->_isNegative = right->_isNegative;
+        result->_exponent = right->_exponent;
+    } else {
+        NSCalculationError error = _subtraction(result, left, right);
+        if (error != NSCalculationNoError) {
+            INVALIDATE_DECIMAL(result);
+            return error;
+        }
+        result->_isNegative = left->_isNegative;
+        result->_exponent = left->_exponent;
+    }
+
+    NSDecimalCompact(result);
+    return NSCalculationNoError;
+}
+
+static NSCalculationError _subtraction(NSDecimal* result, const NSDecimal* left, const NSDecimal* right) {
+    // Note: All inputs have been normalized via NSDecimalNormalize
+    unsigned int commonLength = std::min(left->_length, right->_length);
+
+    unsigned int overflow = 1;
```

Comment:
okay, I understand where this 1 comes from now, and it's incorrect

The correct amount to add at each place is actually (USHRT_MAX + 1), not USHRT_MAX. However, since in a lot of your test cases your first mantissa short tends to 'borrow' from your second short and so on (think doing subtraction by hand, you subtract 1 from the higher place and add 10 to the lower), you end up needing only USHRT_MAX in most of your test cases, making it correct by coincidence.

However, if you do a test case like below where you _don't_ need to borrow:
1 FFFE FFFF
  0001 0001

this algorithm results in
1 FFFC FFFE

instead of
1 FFFD FFFE

---
In reply to: [81235062](https://github.com/Microsoft/WinObjC/pull/1032#discussion_r81235062) [](ancestors = 81235062,81234205)